src/datatype.c

```diff
@@ -824,6 +824,24 @@ JL_DLLEXPORT jl_datatype_t * jl_new_foreign_type(jl_sym_t *name,
     return bt;
 }
 
+JL_DLLEXPORT int jl_reinit_foreign_type(jl_datatype_t *dt,
+                                        jl_markfunc_t markfunc,
+                                        jl_sweepfunc_t sweepfunc)
+{
+    if (!jl_is_foreign_type(dt))
+            return 0;
+    jl_datatype_layout_t *layout = (jl_datatype_layout_t *)
+      jl_gc_perm_alloc(sizeof(jl_datatype_layout_t) + sizeof(jl_fielddescdyn_t),
+        0, 4, 0);
+    *layout = *(dt->layout);
+    jl_fielddescdyn_t * desc =
+      (jl_fielddescdyn_t *) ((char *)layout + sizeof(*layout));
+    desc->markfunc = markfunc;
+    desc->sweepfunc = sweepfunc;
+    dt->layout = layout;
```

Comment:
I think @vtjnash is suggesting effectively this:
```suggestion
    jl_datatype_layout_t *layout = dt->layout;
    jl_fielddescdyn_t * desc =
      (jl_fielddescdyn_t *) ((char *)layout + sizeof(*layout));
    desc->markfunc = markfunc;
    desc->sweepfunc = sweepfunc;
```
Seems fine to me? But perhaps there are hidden pitfalls I am simply not aware of?

Reply:
We need to make sure that when we deserialize a foreing type we actually initialize ith with `+sizeof(jl_fielddescdyn_t)`. I didn't trust us to do so yet.